File: QuasarCode_Library_Maths/Units/Systems.cs
using System;
using System.Collections.Generic;
using System.Text;

using QuasarCode.Library.Maths.Units.Common;

namespace QuasarCode.Library.Maths.Units
{
    /// <summary>
    /// The systems of mesurement with avalable units
    /// </summary>
    public enum Systems
    {
        /// <summary>No present system</summary>
        None,

        /// <summary>SI or Metric</summary>
        SI,

        /// <summary>Imperial (Old)</summary>
        Imperial
    }

    public abstract class SystemBase : ISystem
    {
        public string Name { get; protected set; }
        public Dictionary<Tuple<IQuantity, ISystem, ISystem>, SystemConversion> Conversions { get; protected set; }
        public Dictionary<IQuantity, ISingleUnit> BaseUnits { get; protected set; }

        protected SystemBase(string name, ISingleUnit[] baseUnits, params SystemConversion[] conversions)
        {
            this.Name = name;

            BaseUnits = new Dictionary<IQuantity, ISingleUnit>();
            foreach (ISingleUnit unit in baseUnits)
            {
                BaseUnits.TryAdd(unit.Quantity, unit);
            }

            Conversions = new Dictionary<Tuple<IQuantity, ISystem, ISystem>, SystemConversion>();
            foreach (SystemConversion conversion in conversions)
            {
                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(conversion.ForQuantity, conversion.FromSystem, conversion.ToSystem), conversion);
            }

            if (this.GetType() != typeof(NoneSystem))
            {
                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new NoneQuantity(), this, this),
                    new SystemConversion(new NoneQuantity(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new Angle(), this, this),
                    new SystemConversion(new Angle(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new Length(), this, this),
                    new SystemConversion(new Length(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new Mass(), this, this),
                    new SystemConversion(new Mass(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new Time(), this, this),
                    new SystemConversion(new Time(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new ElectricCurrent(), this, this),
                    new SystemConversion(new ElectricCurrent(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new Temperature(), this, this),
                    new SystemConversion(new Temperature(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new Quantity(), this, this),
                    new SystemConversion(new Quantity(), this, this, (double value) => value));

                Conversions.TryAdd(new Tuple<IQuantity, ISystem, ISystem>(new LuminousIntensity(), this, this),
                    new SystemConversion(new LuminousIntensity(), this, this, (double value) => value));
            }
        }
    }

    public sealed class NoneSystem : SystemBase
    {
        public NoneSystem() : base("None", new ISingleUnit[] { new None() }) { }
    }

    public sealed class SI : SystemBase//TODO: add imperial conversions
    {
        public SI() : base("SI",
            new ISingleUnit[] { new Radian(), new Meter(), new Kilogram(), new Second(), new Ampere(), new Kelvin(), new Moles(), new Candela() }) { }
    }

    public sealed class Imperial : SystemBase//TODO: add imperial units and conversions
    {
        public Imperial() : base("Imperial",
            new ISingleUnit[] { }) { }
    }
}

File: QuasarCode_Python/QuasarCode/Maths/Units/Systems.py
from enum import Enum
from QuasarCode.Maths.Units._ISystem import ISystem
from QuasarCode.Maths.Units.Quantities import NoneQuantity, Angle, Length, Mass, Time, ElectricCurrent, Temperature, Quantity, LuminousIntensity

class Systems(Enum):
    """
    The systems of mesurement with avalable units

    none - No present system
    SI - SI or Metric
    Imperial - Imperial (Old)
    """

    # No present system
    none,

    # SI or Metric
    SI,

    # Imperial (Old)
    Imperial


class SystemBase(ISystem):
    def __init__(self, name, baseUnits, *conversions):
        self.__Name = name

        self.__BaseUnits = {}
        for unit in baseUnits:
            self.__BaseUnits[unit.getQuantity()] = unit

        self.__Conversions = {}
        for conversion in conversions:
            self.__Conversions[(conversion.getForQuantity(), conversion.getFromSystem(), conversion.getToSystem())] = conversion

        if issubclass(self, NoneSystem):
            self.__Conversions[(NoneQuantity(), self, self)] = SystemConversion(NoneQuantity(), this, this, lambda value: value)
            self.__Conversions[(Angle(), self, self)] = SystemConversion(Angle(), this, this, lambda value: value)
            self.__Conversions[(Length(), self, self)] = SystemConversion(Length(), this, this, lambda value: value)
            self.__Conversions[(Mass(), self, self)] = SystemConversion(Mass(), this, this, lambda value: value)
            self.__Conversions[(Time(), self, self)] = SystemConversion(Time(), this, this, lambda value: value)
            self.__Conversions[(ElectricCurrent(), self, self)] = SystemConversion(ElectricCurrent(), this, this, lambda value: value)
            self.__Conversions[(Temperature(), self, self)] = SystemConversion(Temperature(), this, this, lambda value: value)
            self.__Conversions[(Quantity(), self, self)] = SystemConversion(Quantity(), this, this, lambda value: value)
            self.__Conversions[(LuminousIntensity(), self, self)] = SystemConversion(LuminousIntensity(), this, this, lambda value: value)

    def getName(self):
        return self.__Name

    def getConversions(self):
        return self.__Conversions

    def getBaseUnits(self):
        return self.__BaseUnits

class NoneSystem(SystemBase):
    def __init__(self):
        super().__init__("None", [none()])

class SI(SystemBase):
    def __init__(self):
        super().__init__("SI", [Radian(), Meter(), Kilogram(), Second(), Ampere(), Kelvin(), Moles(), Candela()])

class Imperial(SystemBase):
    def __init__(self):
        super().__init__("Imperial", [])